File: docs/bend2/examples/lang-core-effects.js
// JS host half of Core.double, served by `bend ...` (interpreted) and
// `bend ... -o <file>.js`. The compiler wraps this file in a closure and finds
// the function by name: the def name lowercased, dots to underscores. A U32
// arrives as a number and a U32 answer returns as a number.
function core_double(n) {
  return (n * 2) >>> 0;
}

File: docs/bend2/examples/lang-core-effects.c
// C host half of Core.double, served by `bend ... -o <binary>`.
// The compiler splices this file into the program's C source after the
// runtime, so runtime symbols (Term, Env, IoWork, io_eff, term words) are in
// scope. The effect is pure argument marshaling: a U32 arrives as the word
// f[0]; a U32 answer is returned as a Term.
Term core_double_run(Env e, Term* f, IoWork* w) {
  return (Term)((u32)f[0] * 2u);
}

// Registration: CID_CORE_DOUBLE is the def name uppercased, dots to
// underscores; need 0 runs the effect at once on the event loop.
static void __attribute__((constructor)) core_double_use(void) {
  io_eff(CID_CORE_DOUBLE, core_double_run, 0);
}
